optimization/sggp/sggp-algorithm-with-test.metta

```diff
@@ -0,0 +1,597 @@
+)
+(= (punishProd ($prod1 $prod2) $learnRate $cardinality $stocSpace)
+    (let* (
+        ($res (punsihProd $prod1 $learnRate $cardinality $stocSpace))
+        ($res' (punishProd $prod2 $learnRate $cardinality $stocSpace))
+    )
+        ()
+    )
+)
+
+;; This procedure implements the reinforcement of the reinforcement of a grammar given a Tree
+(= (reinforceGrammarbasedOnTree $tree $learnRate $cardinality $grammarSpace $stocSpace)
+    (let* (
+        ($relations (getRelations $tree ()))
+        ($prodRules (collapse (getRuleFromSubTree (superpose $relations) $stocSpace)))
+        ;; (() (println! (production rules $prodRules)))
+        ;; ($reinforcedOutput (reinforceTreeHelper $prodRules $learnRate $cardinality $stocSpace))
+        
+    )
+        (reinforceTreeHelper $prodRules $learnRate $cardinality $stocSpace)
+        
+        ;;(collapse (reinforceProd (superpose $prodRules) $learnRate $cardinality $stocSpace))
+        ;; $reinforcedOutput
+    )
+)
+
+;; The below piece of code punishes the production rule in a stochastic grammar based on a tree
+(= (punishGrammarbasedOnTree $tree $learnRate $cardinality $grammarSpace $stocSpace)
+    (let* (
+        ($relations (getRelations $tree ()))
+        ($prodRules (collapse (getRuleFromSubTree (superpose $relations) $stocSpace)))
+        ;; (println (production rules $prodRules))
+    )
+        ;;(collapse (punishProd (superpose $prodRules) $learnRate $cardinality $stocSpace))
+        (punishTreeHelper $prodRules $learnRate $cardinality $stocSpace)
+    )
+)
+;;This function generates trees guided by the stochastic grammar
+;; algorithm for generating instances // Exploitation purposes
+;; step 1 start from an init symbol with a depth = n.
+;; match the grammar atomspace to get matching tokens as a right handside from the context free grammar
+;; select probabilistically the production rule that passes the threshold
+;; build the tree recursively and non-deterministically. with depth (n-1)
+;; and make the rhs to lhs and start step 2
+;; Does this procedure result in a well formed tree?
+;; Do this until the depth is n = 0
+;; after finishing collapse the non-deterministic result.
+;; Question, Does this result in a well formed tree?
+;; Definition: A well formed tree is a recursive data structure were internal nodes have atleast two children 
+;; This stands out if the internal node is not a NOT junctor.
+;; If the internal node is a NOT junctor it will have a one child
+
+(= (generateTrees $initSymbol $grammarSpace $stocSpace $explorationCoeff $depth)
+    (if (== $initSymbol start)
+        (generateTrees expr $grammarSpace $stocSpace $explorationCoeff $depth)
+        (let $derivs (findSuitableProdRules $initSymbol $stocSpace)
+            (if (== $derivs ())
+                $initSymbol
+                (if (<= $depth 0)
+                    ;; (match $grammarSpace ($initSymbol $x) (if (== $x ()) $initSymbol $x))
```

Comment:
Why is this code commented out? Is there a case where this commented out code would be usefull? If so, add that case as a comment so that it won't be forgotten in the future. If it won't be useful remove it.